src/VisualStudio/Core/Def/ProjectSystem/BrokeredService/WorkspaceProject.cs

```diff
@@ -159,9 +159,32 @@ public Task SetProjectHasAllInformationAsync(bool hasAllInformation, Cancellatio
             return Task.CompletedTask;
         }
 
-        public Task<IAsyncDisposable> StartBatchAsync(CancellationToken cancellationToken)
+        public Task<IWorkspaceProjectBatch> StartBatchAsync(CancellationToken cancellationToken)
         {
-            return Task.FromResult(_project.CreateBatchScope());
+            return Task.FromResult<IWorkspaceProjectBatch>(new WorkspaceProjectBatch(_project.CreateBatchScope()));
+        }
+
+        private class WorkspaceProjectBatch : IWorkspaceProjectBatch
+        {
+            private IAsyncDisposable? _batch;
+
+            public WorkspaceProjectBatch(IAsyncDisposable batch)
+            {
+                _batch = batch;
+            }
+
+            public async Task ApplyAsync(CancellationToken cancellationToken)
+            {
+                if (_batch == null)
```

Comment:
nit: interlocked exchange this.

Reply:
There's no expectation of parallel use of the interface, generally -- this is fine.